Portal/src/Datahub.Functions/ProjectUsageUpdater.cs

```diff
@@ -11,38 +12,69 @@ namespace Datahub.Functions;
         CancellationToken cancellationToken)
     {
+        var rolledOver = false;
         // test for ping
         if (await _pongService.Pong(queueItem))
-            return;
+            return false;
 
         // deserialize message
         var message = DeserializeQueueMessage(queueItem);
 
-        // get resource groups
-        var resourceGroups = EnumerateResourceGroups(message).ToArray();
+        _logger.LogInformation("Querying cost management...");
+        var (costRollover, spentAmount) =
+            await _workspaceCostMgmtService.UpdateWorkspaceCostAsync(message.SubscriptionCosts, message.ProjectAcronym);
+        _logger.LogInformation("Querying budget...");
+        var budgetSpentAmount = await _workspaceBudgetMgmtService.UpdateWorkspaceBudgetSpentAsync(message.ProjectAcronym);
 
-        // update the usage
-        if (!await _usageService.UpdateProjectUsage(message.ProjectId, resourceGroups, cancellationToken))
-            return;
+        // The query to cost checks if the last update was outside of the current fiscal year, if so that means we are in a new fiscal year
+        // The query to budget checks if the amount spent captured by the budget is less than previously. If so, that means the budget was renewed.
+        if (message.ForceRollover || costRollover)
+        {
+            _logger.LogInformation($"Budget rollover initiated.");
+            var currentBudget = await _workspaceBudgetMgmtService.GetWorkspaceBudgetAmountAsync(message.ProjectAcronym);
+            _logger.LogInformation($"Spend captured by cost management: {spentAmount}");
+            _logger.LogInformation($"Spend captured by budget : {budgetSpentAmount}");
+            
+            // Checking if the two captured costs are within 5% of each other to ensure that the rollover is valid
+            var relativeDifference = (budgetSpentAmount - spentAmount) / budgetSpentAmount;
+            if ( relativeDifference <= (decimal)0.05)
+            {
+                _logger.LogInformation($"Executing rollover for {message.ProjectAcronym}");
+                await _workspaceBudgetMgmtService.SetWorkspaceBudgetAmountAsync(message.ProjectAcronym,
+                    currentBudget - budgetSpentAmount, true);
+                rolledOver = true;
+                // Generate fiscal year report here?
+            }
+            else
+            {
+                _logger.LogWarning($"Aborted rollover due to large difference between captured spend and captured costs ({relativeDifference:P})");
+            }
+        }
 
         // queue the usage notification message
-        await _mediator.Send(new ProjectUsageNotificationMessage(message.ProjectId), cancellationToken);
+        await _mediator.Send(new ProjectUsageNotificationMessage(message.ProjectAcronym), cancellationToken);
+        return rolledOver;
```

Comment:
What does the `bool` return value do for a queue triggered function? Is that the status?

Reply:
I don't believe it does anything functional. This was for testing purposes: the specflow tests hit this method directly and use the return value as part of assertions